### PR TITLE
[AzureMonitorLiveMetrics] POC

### PR DESCRIPTION
**Pull request by TimothyMothra**
This PR commits a working Proof of Concept. This is the first working POC that gets data into the Azure Portal. 
THIS IS NOT FINAL. After committing these changes, I will merge my POC with Raj's and work on additional enhancements.

image

## Changes
- Demo project
  - updated Program.cs to generate telemetry in a loop. This is needed to generate traffic to see LiveMetrics in the Azure Portal.
- LiveMetrics project
  - Added a Customization of `QuickPulseSDKClientAPIsRestClient.cs`. This is needed because the Generated RestClient has a deserialization bug.
  - Added `ManagerFactory`, `Manager`, and `Manager.Metrics`. 
    This is modeled off of the Transmitter class in our AzureMonitorExporter. The Factory ensures that we only have one instance of the Manager per ConnectionString. The Manager class has the HTTP traffic (Ping & Post). The Manager.Metrics handles our metrics.
  - Added `LiveMetricsMetricExporter` to translate OTel Metrics to AzureMonitor Metrics
  - Added `LiveMetricsTraceExporter` as the entry point to my POC and is used to initialize the Manager class.

## Future Work
- My Exporters will be merged with Raj's
- Consolidate Constants.
- Replace OTel Metrics SDK with custom implementation for lower overhead.

### CODE REVIEW

File: sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/Azure.Monitor.OpenTelemetry.LiveMetrics.csproj [hunk 1]
```diff
@@ -14,6 +14,7 @@
     <PackageReference Include="Azure.Core" />
     <PackageReference Include="OpenTelemetry" />
     <PackageReference Include="OpenTelemetry.Exporter.Console" VersionOverride="1.6.0" />
+    <PackageReference Include="System.Diagnostics.PerformanceCounter" VersionOverride="7.0.0" />
```
Comment (TimothyMothra) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/Azure.Monitor.OpenTelemetry.LiveMetrics.csproj:17]: This is needed to read Performance Counters (CPU & Memory) to enable POC. Will work on removing this in the future.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/Customizations/QuickPulseSDKClientAPIsRestClient.cs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            switch (message.Response.Status)
+            {
+                case 200:
+                    {
+                        CollectionConfigurationInfo value = default;
+                        if (message.Response.Headers.ContentLength != 0)
+                        {
+                            using var document = JsonDocument.Parse(message.Response.ContentStream);
+                            value = CollectionConfigurationInfo.DeserializeCollectionConfigurationInfo(document.RootElement);
+                        }
```
Comment (TimothyMothra) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/Customizations/QuickPulseSDKClientAPIsRestClient.cs:49]: Bug fix. Generated code doesn't check that ContentLength > 0 before deserializing.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/Azure.Monitor.OpenTelemetry.LiveMetrics.csproj (hunk 1 above)
Comment (rajkumar-rangaraj) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/Azure.Monitor.OpenTelemetry.LiveMetrics.csproj:17]: We may to keep this one forever, there is no good way to get this value.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/LiveMetricsExporterOptions.cs [hunk 3]
```diff
@@ -10,7 +10,7 @@ namespace Azure.Monitor.OpenTelemetry.LiveMetrics;
 /// <summary>
 /// Options that allow users to configure the Live Metrics.
 /// </summary>
-internal class LiveMetricsExporterOptions : ClientOptions
+public class LiveMetricsExporterOptions : ClientOptions
```
Comment (TimothyMothra) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/LiveMetricsExporterOptions.cs:13]: Needed this public for Demo project

File: sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/LiveMetricsExtensions.cs [hunk 4]
```diff
@@ -14,7 +14,7 @@ namespace Azure.Monitor.OpenTelemetry.LiveMetrics
     /// <summary>
     /// Extension methods to register Live Metrics.
     /// </summary>
-    internal static class LiveMetricsExtensions
+    public static class LiveMetricsExtensions
```
Comment (TimothyMothra) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/LiveMetricsExtensions.cs:17]: Needed this public for Demo project

File: sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/LiveMetricsExtractionProcessor.cs [hunk 5]
```diff
@@ -98,11 +106,11 @@ public override void OnEnd(Activity activity)
...
                 if (IsSuccess(activity, statusCodeAttributeValue))
                 {
-                    _requestFailedPerSecond.Add(1);
+                    _requestSucceededPerSecond.Add(1);
                 }
                 else
                 {
-                    _requestSucceededPerSecond.Add(1);
+                    _requestFailedPerSecond.Add(1);
                 }
```
Comment (TimothyMothra) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/LiveMetricsExtractionProcessor.cs:114]: bug fix. :)

File: sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/Azure.Monitor.OpenTelemetry.LiveMetrics.csproj (hunk 1 above)
Comment (TimothyMothra) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/Azure.Monitor.OpenTelemetry.LiveMetrics.csproj:17]: I'm fine with this dependency, as System.Diagnostic is a core dependency of OTel.
Application Insights is able to read perf counters without an extra dependency. I need more time to investigate how that works. For short term, this is fine. :)

Comment (rajkumar-rangaraj) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/Azure.Monitor.OpenTelemetry.LiveMetrics.csproj:17]: AI SDK uses  `System.Diagnostics.PerformanceCounter` - https://github.com/microsoft/ApplicationInsights-dotnet/blob/de66d679ff32f5a74553edbf52b10b9dc57ded70/WEB/Src/PerformanceCollector/PerformanceCollector/Perf.csproj#L32

File: sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/api/Azure.Monitor.OpenTelemetry.LiveMetrics.netstandard2.0.cs [hunk 6]
```diff
@@ -1,3 +1,22 @@
+namespace Azure.Monitor.OpenTelemetry.LiveMetrics
+{
+    public partial class LiveMetricsExporterOptions : Azure.Core.ClientOptions
+    {
+        public LiveMetricsExporterOptions() { }
+        public string ConnectionString { get { throw null; } set { } }
+        public Azure.Core.TokenCredential Credential { get { throw null; } set { } }
```
Comment (rajkumar-rangaraj) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/api/Azure.Monitor.OpenTelemetry.LiveMetrics.netstandard2.0.cs:7]: Does live metrics support AAD? If not remove changes related to it.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/api/Azure.Monitor.OpenTelemetry.LiveMetrics.netstandard2.0.cs [hunk 7]
```diff
@@ -1,3 +1,22 @@
+namespace Azure.Monitor.OpenTelemetry.LiveMetrics
+{
+    public partial class LiveMetricsExporterOptions : Azure.Core.ClientOptions
```
Comment (rajkumar-rangaraj) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/api/Azure.Monitor.OpenTelemetry.LiveMetrics.netstandard2.0.cs:3]: Do we need to need to inherit ClientOptions?

Comment (rajkumar-rangaraj) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/api/Azure.Monitor.OpenTelemetry.LiveMetrics.netstandard2.0.cs:3]: We could validate later too.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/api/Azure.Monitor.OpenTelemetry.LiveMetrics.netstandard2.0.cs (hunk 6 above)
Comment (TimothyMothra) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/api/Azure.Monitor.OpenTelemetry.LiveMetrics.netstandard2.0.cs:7]: Yes it does.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/api/Azure.Monitor.OpenTelemetry.LiveMetrics.netstandard2.0.cs (hunk 7 above)
Comment (TimothyMothra) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/api/Azure.Monitor.OpenTelemetry.LiveMetrics.netstandard2.0.cs:3]: Yes, this is required to initialize the RestClient.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/LiveMetricsTraceExporter.cs [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+using System.Diagnostics;
+using Azure.Monitor.OpenTelemetry.LiveMetrics.Internals;
+using OpenTelemetry;
+
+namespace Azure.Monitor.OpenTelemetry.LiveMetrics
+{
+    /// <summary>
+    /// Empty Exporter. This is used to enable the Manager Singleton.
+    /// </summary>
+    public class LiveMetricsTraceExporter : BaseExporter<Activity>
```
Comment (rajkumar-rangaraj) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/LiveMetricsTraceExporter.cs:13]: Why this is needed?

Comment (TimothyMothra) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/LiveMetricsTraceExporter.cs:13]: This is the entry point for my POC. Adding this to a TracerProvider initializes the Manager Singleton.

**Review by rajkumar-rangaraj — APPROVED** [review]
Approving as POC. This requires a complete clean up later.

**Pull request merged**